locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/BasicLocationNetworkInfoCustomizer.java

```diff
@@ -0,0 +1,473 @@
+
+        if (credChoice == null) {
+            credChoice = Iterables.getFirst(credentialCandidates, null);
+            if (credChoice == null) {
+                throw new IllegalStateException("Exhausted all options when determining credential for " + location);
+            }
+        }
+
+        if (contextEntity != null) {
+            contextEntity.sensors().set(Attributes.ADDRESS, hapChoice.getHostText());
+        }
+        ManagementAddressResolveResult result = new ManagementAddressResolveResult(hapChoice, credChoice);
+        LOG.debug("{} resolved management parameters for {} in {}: {}",
+                new Object[]{this, location, Duration.of(timer), result});
+        return result;
+    }
+
+    private boolean shouldPublishNetworks() {
+        return Boolean.TRUE.equals(config().get(PUBLISH_NETWORKS));
+    }
+
+    // TODO: Separate this into second part?
+    void publishNetworks(NodeMetadata node, Entity entity) {
+        // todo hostnames?
+        int i = 0;
+        for (String address : node.getPrivateAddresses()) {
+            final AttributeSensor<String> sensor = Sensors.newStringSensor("host.address.private." + i++);
+            if (entity.sensors().get(sensor) == null) {
+                entity.sensors().set(sensor, address);
+            }
+        }
+        i = 0;
+        for (String address : node.getPublicAddresses()) {
+            final AttributeSensor<String> sensor = Sensors.newStringSensor("host.address.public." + i++);
+            if (entity.sensors().get(sensor) == null) {
+                entity.sensors().set(sensor, address);
+            }
+        }
+    }
+
+    // --------------------------------------------------------------------------------------
+
+    /**
+     * Returns the hosts and ports that should be considered when determining the address
+     * to use when connecting to the location by assessing the following criteria:
+     * <ol>
+     *     <li>Use the hostAndPortOverride set in options.</li>
+     *     <li>If the machine is connectable, user credentials are given and the machine is provisioned
+     *     in AWS then use {@link JcloudsLocation#getHostnameAws(NodeMetadata, Optional, Supplier, ConfigBag)}.</li>
+     *     <li>If the machine is connectable and pollForFirstReachableAddress is set in options then use all
+     *     {@link #getReachableAddresses reachable} addresses.</li>
+     *     <li>Use the first address that is resolvable with {@link #isAddressResolvable}.</li>
+     *     <li>Use the first address in the node's public then private addresses.</li>
+     * </ol>
+     */
+    protected Iterable<HostAndPort> getManagementCandidates(
+            JcloudsLocation location, NodeMetadata node, ConfigBag config, ManagementAddressResolveOptions options) {
+        final Optional<HostAndPort> hostAndPortOverride = options.getHostAndPortOverride();
+        boolean lookupAwsHostname = Boolean.TRUE.equals(config.get(JcloudsLocation.LOOKUP_AWS_HOSTNAME));
```

Comment:
It would be nice to factor out the AWS specifics into a separate tool; I know this has come from the existing JcloudsLocation code but as this is a new class maybe there is an opportunity to avoid repeating this pattern?